### ISSUE TITLE
Eval: `EvalPackage` has no runner — add a trials engine (K isolated attempts per case, unified scorer contract, full-trajectory records, pass-rate + frontier selection)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAI#3369]
## Aim

Answer the question "**does my agent actually work?**" with statistics instead of one sampled run: take a defined set of task cases, run the agent **K independent times per case** (each attempt isolated), **score every attempt** through one pluggable scorer contract, and report **per-case pass rates** — including the **frontier band** (cases with pass rate strictly between 0 and 1), which is where an agent is capable-but-inconsistent and where improvement effort (and, via the sibling issue, training-data export) pays off. This is verification as a first-class SDK capability — the measurement half of an improvement pipeline whose data-generation half is the sibling `praisonai-train` issue.

Everything proposed extends what already exists in `praisonaiagents/eval/` — no new subsystem. Validated against current `main` (`309cb97`).

## Current behaviour (every stage exists in isolation; no joints)

**1. Cases exist; the runner is a bare Protocol.** `EvalCase` (name/input/expected/criteria/metadata/timeout — `eval/package.py:7-52`), `EvalPackage` (cases + thresholds + seed, serialisable — `package.py:126-181`), `EvalResult`/`EvalReport` (`package.py:56-122`) are all real. But `EvalRunnerProtocol` (`package.py:184-203`) has **no implementation anywhere in `src/`** — the only references are unit tests. A user can declare a case dataset; nothing iterates it against an agent.

**2. K independent attempts don't exist.** No `n_attempts`/`pass@k`/variance anywhere in `eval/` or `praisonai-train`. The closest pieces are single-case judge loops: `AccuracyEvaluator(num_iterations=N)` (`eval/accuracy.py:35,157-178` — one input, N judge scores, no per-attempt verdict) and `CriteriaEvaluator(num_iterations=N)` (`criteria.py:38,197-225`). `EvaluationLoop` is sequential refinement, not independent samples (`loop.py:186-228`); `praisonai-train`'s `AgentTrainer(iterations=3)` deliberately chains attempts by injecting prior feedback (`train/agents/orchestrator.py:301-376`). The one real case-runner (`praisonai-code` `csv_test_runner`) parses a `retries` field and **never reads it** (`csv_test_runner.py:43,67` — dead field). `EvalSuite._run_parallel` is still a warn-and-fallback stub (`eval/suite.py:183-188`).

**3. Scoring has no unified contract.** Verdicts are LLM-judge-centric (`eval/judge.py`, `grader.py`); deterministic verifiers exist but are siloed one-offs — `ReliabilityEvaluator` (tool assertions, `reliability.py:20-56`), `HarnessEvaluator` (4 boolean gates over an externally-produced trace dict, `harness_eval.py:160-224`, **zero non-test callers**), `ContextEvaluator` (`context_eval.py:169-298`). Exactly **one** numeric verifier callable exists — `PromptOptimizer(metric=...)` (`prompt_optimizer.py:91,118-131`) — and it drives candidate selection, not per-attempt pass/fail. (Adjacent bug: `prompt_optimizer.py:38` defines a local `EvalCase = Tuple[str, Any]` that shadows `package.EvalCase` — a real `EvalCase` will not work there.)

**4. Attempt records carry no trajectory.** Every eval result stores a final string only: `EvaluationScore` (`results.py:15-26`), `EvalResult.actual_output: Optional[str]` (`package.py:56-64`); `HarnessEvaluator.to_eval_result` even sets `actual_output=None` (`harness_eval.py:238`). Meanwhile rich trajectory capture **already exists, unconnected**: `trace/context_events.py` (`ContextTraceEmitter` — `message_added:496`, `tool_call_start/end:525,545`, `llm_request/response:668,704`), `trace/protocol.py` (`TraceEmitter`, JSONL sink), and the unwired `runtime/journal.py` `RunJournal` (constructed only in its own docstring and tests).

## Desired behaviour

```python
from praisonaiagents.eval import EvalPackage, EvalCase, run_trials

package = EvalPackage(name="checkout-flows", cases=[
    EvalCase(name="refund", input="...", expected="...",
             verify=my_metric_fn),                     # NEW optional field — see scorer contract
    EvalCase(name="lookup", input="...", criteria=["cites the order id"]),   # judge-scored, as today
])

report = run_trials(agent, package, k=8, concurrency=4)
print(report.pass_rates())        # {"refund": 0.75, "lookup": 1.0}
print(report.frontier())          # cases with 0 < pass_rate < 1 → where to focus
report.save("trials.json")        # full attempt records, reloadable
```

1. **`run_trials(agent, package, *, k=1, concurrency=1)`** — the missing `EvalRunnerProtocol` implementation. K independent attempts per case; async fan-out with a semaphore; failures recorded as data (`stop_reason`: completed/error/timeout), never exceptions.
2. **Per-attempt isolation, unconditional:** each attempt runs on a fresh copy of the agent (config copy + fresh session id, response caches disabled, memory/knowledge **writes** severed for the attempt) so K attempts are K samples, not one replay and not cross-contamination.
3. **One scorer contract:** `TrialScore(value: float∈[0,1], passed: bool, reason: str|None)`. Sources unify behind it: an `EvalCase.verify` callable `(output_or_record, expected) -> bool|float|TrialScore`; the existing `Judge`/criteria path; and the existing tool-assertion logic (adapt `ReliabilityEvaluator`). **Unscored ≠ zero**: a timed-out/errored attempt has `score=None` and is excluded from statistics, never coerced to 0.
4. **Trajectory-carrying attempt records:** each attempt captures the messages and tool executions (reuse `ContextTraceEmitter`/`TraceEmitter` — attach a per-attempt sink) so the record is reusable downstream (the training-export sibling depends on this). `EvalResult` gains an optional `record` field; default persistence includes it.
5. **Aggregation:** per-case `n_scored/n_passed/pass_rate/mean_value`, package-level totals, `frontier()` (0 < pass_rate < 1), and a stable `summary()` dict for CI gating.
6. Fix the two adjacent defects en route: the `prompt_optimizer.py:38` `EvalCase` shadowing, and either implement or remove the dead `retries` field / `_run_parallel` stub.

Out of scope (deliberately, to stay lean): seeds/temperature sweeps, pass@k math beyond plain pass-rate, cross-run diffing, any training call — the export half lives in the sibling `praisonai-train` issue.

## Layer placement

- **Primary layer:** core (`praisonaiagents/eval`) — extends `EvalPackage`, `Judge`, `ReliabilityEvaluator`, `trace/`; stdlib + existing deps only (asyncio, copy), honouring the protocol-first/no-heavy-imports rule.
- **Why not wrapper:** nothing here is CLI/config-specific; a `praisonai eval trials` CLI is a thin optional follow-up.
- **Why not `praisonai-train`:** measurement must not depend on training deps; train consumes the report (sibling issue), not the other way round.
- **Why not tools/plugins:** this is the eval harness itself, not an agent-callable tool or lifecycle hook.
- **3-way surface:** Python now; CLI/YAML follow-ups once the engine exists.

## Implementation plan (current-main anchors)

1. **`eval/trials.py` (new):** `TrialAttempt` (record, score, stop_reason, duration), `TrialReport` (cases → attempts, `pass_rates()/frontier()/summary()/save()/load()`), `run_trials`/`arun_trials` implementing `EvalRunnerProtocol` (`package.py:184-203`).
2. **Scorer contract:** `TrialScore` dataclass in `eval/base.py`; resolution order per case — `verify` callable → tool assertions (from `EvalCase.metadata` expected_tools, reusing `reliability.py:20-56` logic) → `criteria`/judge (existing `Judge`); scorer runs only on completed attempts.
3. **Isolation helper:** `_isolated_agent_copy(agent)` — copy config, fresh session id, disable response caching, sever memory/knowledge write flags for the attempt (fields exist on `Agent`; enumerate and pin with a drift test).
4. **Trajectory capture:** per-attempt `ListSink` on the existing emitters (`trace/protocol.py:139-167`, `trace/context_events.py`); store as `TrialAttempt.record` (messages + tool calls), serialised in `save()`.
5. **`EvalCase.verify: Optional[Callable]`** added to `package.py:7-52` (backward compatible — default None keeps judge path).
6. Fix `prompt_optimizer.py:38` shadowing (rename local alias); remove dead `retries` (`csv_test_runner.py:67`) or wire it to `k`.

## Tests (TDD)

- `test_run_trials_k_attempts_per_case` — K×N attempts, deterministic assembly order.
- `test_attempt_isolation` — a memory-writing agent leaves caller memory untouched; two attempts share no session id.
- `test_verify_callable_bool_float_score` — all three return shapes normalise to `TrialScore`.
- `test_unscored_excluded_from_stats` — timeout attempt → `score=None`, pass_rate denominator unaffected.
- `test_frontier_selection` — only 0<rate<1 cases returned.
- `test_judge_path_unchanged` — criteria-only case scores via existing Judge exactly as `CriteriaEvaluator` would.
- `test_report_save_load_roundtrip` — records + scores survive.
- Real agentic test: a 2-case package, k=4, real LLM — report shows plausible pass rates and each attempt carries a non-empty message record.

## Severity

**High** — verification-by-statistics is the difference between "it worked once in a demo" and "it works"; the SDK already ships the cases, the judges, and the trace emitters, so the missing piece is precisely the joints. It is also the hard prerequisite for the training-data pipeline (sibling issue) and for the existing `PromptOptimizer` to optimise against real pass rates.

## Validation (current main `309cb97`)

- Runner absent: `EvalRunnerProtocol` implementations — none outside tests (`package.py:184-203`).
- K-attempts absent: no `n_attempts/pass@k` hits; `accuracy.py:157-178`/`criteria.py:197-225` are single-case; dead `retries` (`csv_test_runner.py:43,67`); parallel stub (`suite.py:183-188`).
- Scoring siloed: single metric callable (`prompt_optimizer.py:91`); shadowing bug (`:38`); deterministic verifiers unconnected (`reliability.py:20-56`, `harness_eval.py:160-224` zero callers).
- Trajectories not in results: `results.py:15-26`, `package.py:56-64`, `harness_eval.py:238`; capture exists unconnected (`trace/context_events.py:496-704`, `runtime/journal.py` unwired).

## Related

- Sibling: `praisonai-train` export of passing trial records as a fine-tuning dataset (consumes `TrialReport`).
- #3104 (`PromptOptimizer`) — once trials exist, its metric path can optimise against real pass rates.
